### PR TITLE
Make pd.Period immutable

### PR DESCRIPTION
**Pull request by jbrockmendel**
Follows @jreback 's suggestion in #17116 to define `_Period.ordinal` and `_Period.freq` as `def readonly` attributes.  This requires massaging `__new__/__init__` a bit.

Adds a quick test that these attributes can't be changed.

Note that this doesn't fully solve the issue of Periods being immutable because `per.freq` is a `DateOffset` object which itself can be messed with.

- [ ] closes #
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/test_lib.py [hunk 1]
```diff
@@ -248,3 +248,13 @@ def test_empty_like(self):
         expected = np.array([True])
 
         self._check_behavior(arr, expected)
+
+
+def test_period_immutable():
+    per = pd.Period('2014Q1')
```
Comment (jreback) [inline pandas/tests/test_lib.py]: add the issue number here. move test to tests for period scalars.

File: pandas/_libs/period.pyx [hunk 2]
```diff
@@ -19,8 +18,11 @@ from pandas import compat
 from pandas.compat import PY2
 
 cimport cython
+# this is _libs.src.datetime, not python stdlib
```
Comment (jreback) [inline pandas/_libs/period.pyx]: don't add superfluous comments

**Review by jreback — CHANGES_REQUESTED** [review]
add a whatsnew note in API breaking changes.

File: pandas/tests/test_lib.py (hunk 1 above)
Comment (jbrockmendel) [inline pandas/tests/test_lib.py]: Sure.

File: pandas/_libs/period.pyx (hunk 2 above)
Comment (jbrockmendel) [inline pandas/_libs/period.pyx]: Similar comments exist elsewhere e.g. in lib.pyx:

```
# this is our tseries.pxd
from datetime cimport *
```

This name overlap is an easy source of confusion.

Not worth arguing about so I'll change it.  Just sayin'.

File: pandas/_libs/period.pyx [hunk 3]
```diff
@@ -718,9 +724,8 @@ cdef class _Period(object):
         if ordinal == iNaT:
             return NaT
         else:
-            self = _Period.__new__(cls)
-            self.ordinal = ordinal
-            self.freq = cls._maybe_convert_freq(freq)
+            freq = cls._maybe_convert_freq(freq)
+            self = _Period.__new__(cls, ordinal, freq)
```
Comment (MaximilianR) [inline pandas/_libs/period.pyx]: Should this be `self = cls.__new__...`?

Comment (jbrockmendel) [inline pandas/_libs/period.pyx]: I don't think so.  `cls._from_ordinal` gets called at the end of `Period.__new__`, so I expect it explicitly needs to be `_Period`.

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -138,6 +138,16 @@ Other Enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. _whatsnew_0210.api_breaking.Period:
+
+pd.Period objects are now immutable
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: just say Period

File: doc/source/whatsnew/v0.21.0.txt [hunk 5]
```diff
@@ -138,6 +138,16 @@ Other Enhancements
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:class:`Period` will now raise an ``AttributeError`` when a user tries to
+assign a new value to the `ordinal` or `freq` attributes.  This will allow
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: you dont need to provide an explanation.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.21.0.txt [hunk 6]
```diff
@@ -138,6 +138,16 @@ Backwards incompatible API changes
...
 
 
+.. _whatsnew_0210.api_breaking.Period:
+
+pd.Period objects are now immutable
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:class:`Period` will now raise an ``AttributeError`` when a user tries to
+assign a new value to the `ordinal` or `freq` attributes (:issue:`17116`).
+=======
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: extra line here. IIRC I told you to make a sub-section, but let's move this back to just a single line (still in API breaking change)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/test_period.py [hunk 7]
```diff
@@ -1406,3 +1406,14 @@ def test_period_ops_offset(self):
 
         with tm.assert_raises_regex(period.IncompatibleFrequency, msg):
             p - offsets.Hour(2)
+
+
+def test_period_immutable():
+    # see gh-17116
+    per = pd.Period('2014Q1')
+    with pytest.raises(AttributeError, message="is not writable"):
```
Comment (jorisvandenbossche) [inline pandas/tests/scalar/test_period.py]: the `message="is not writable"` is not needed here, the default assertion message of pytest is normally clear.

If you actually want to check the message of the AttributeError, it is the `match` keyword you need to use, not `message` (but I don't think it is needed to test that, it is not a custom error message we write ourselves)

Comment (jbrockmendel) [inline pandas/tests/scalar/test_period.py]: Pushing change now.

File: doc/source/whatsnew/v0.21.0.txt (hunk 6 above)
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.21.0.txt]: OK.  RST is another area where I'm figuring it out as I go.  About to push a change that _I think_ corresponds to what you're saying.

File: doc/source/whatsnew/v0.21.0.txt [hunk 8]
```diff
@@ -138,6 +138,14 @@ Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 
+.. _whatsnew_0210.api_breaking.Period:
+
+pd.Period objects are now immutable
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:class:`Period` will now raise an ``AttributeError`` when a user tries to assign a new value to the `ordinal` or `freq` attributes (:issue:`17116`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: no what I meant was remove this as a sub-section and just add it as a single like in other breaking changes.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.21.0.txt (hunk 8 above)
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.21.0.txt]: All the other breaking changes have a subsection like:

```
.. _whatsnew_0210.api_breaking.pandas_eval:

Improved error handling during item assignment in pd.eval
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I'll keep pushing guesses until the Infinite Monkeys Theorem kicks in.

Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.21.0.txt]: Still not fully what he meant :-) 

The "Backwards incompatible API changes" section starts with subsections for things that we think are important enough to explain in more than a single line, and at the end there is section "other API changes" with a bunch of single entries:

https://github.com/jbrockmendel/pandas/blob/19567c8f0de212dec15e17dc918e6eada6a8c354/doc/source/whatsnew/v0.21.0.txt#L280-L295

So you can put it there.

**Review by jorisvandenbossche — APPROVED** [review]
apart from the small move in whatsnew, +1 to merge

**Comment by jreback** [comment]
thanks!

**Pull request merged**